extensions/notebook/src/test/common/utils.test.ts

```diff
@@ -134,7 +137,169 @@ describe('Utils Tests', function () {
+			serverMinorVersion: -1,
+			serverReleaseVersion: -1,
+			engineEditionId: -1,
+			serverVersion: '',
+			serverLevel: '',
+			serverEdition: '',
+			isCloud: false,
+			azureVersion: -1,
+			osVersion: '',
+			options: {}
+		};
+		it('empty endpoints does not error', () => {
+			const serverInfo = Object.assign({}, baseServerInfo);
+			serverInfo.options['clusterEndpoints'] = [];
+			should(utils.getClusterEndpoints(serverInfo).length).equal(0);
+		});
+
+		it('endpoints without endpoint field are created successfully', () => {
+			const serverInfo = Object.assign({}, baseServerInfo);
+			const ipAddress = 'localhost';
+			const port = '123';
+			serverInfo.options['clusterEndpoints'] = [{ ipAddress: ipAddress, port: port }];
+			const endpoints = utils.getClusterEndpoints(serverInfo);
+			should(endpoints.length).equal(1);
+			should(endpoints[0].endpoint).equal('https://localhost:123');
+		});
 
+		it('endpoints with endpoint field are created successfully', () => {
+			const endpoint = 'https://myActualEndpoint:8080';
+			const serverInfo = Object.assign({}, baseServerInfo);
+			serverInfo.options['clusterEndpoints'] = [{ endpoint: endpoint, ipAddress: 'localhost', port: '123' }];
+			const endpoints = utils.getClusterEndpoints(serverInfo);
+			should(endpoints.length).equal(1);
+			should(endpoints[0].endpoint).equal(endpoint);
+		});
+	});
+
+	describe('getHostAndPortFromEndpoint', () => {
+		it('valid endpoint is parsed correctly', () => {
+			const host = 'localhost';
+			const port = '123';
+			const hostAndIp = utils.getHostAndPortFromEndpoint(`https://${host}:${123}`);
+			should(hostAndIp).deepEqual({ host: host, port: port });
+		});
+		it('invalid endpoint is returned as is', () => {
+			const host = 'localhost';
+			const hostAndIp = utils.getHostAndPortFromEndpoint(`https://${host}`);
+			should(hostAndIp).deepEqual({ host: host, port: undefined });
+		});
+	});
+
+	describe('exists', () => {
+		it('runs as expected', async () => {
+			const filename = path.join(os.tmpdir(), `NotebookUtilsTest_uuid.v4()`);
+			try {
+				should(await utils.exists(filename)).be.false();
+				await fs.writeFile(filename, '');
+				should(await utils.exists(filename)).be.true();
+			} finally {
+				try {
+					await fs.unlink(filename);
+				} catch { /* no-op */ }
+			}
+		});
+	});
+
+	describe('getIgnoreSslVerificationConfigSetting', () => {
+		it('runs as expected', async () => {
+			should(utils.getIgnoreSslVerificationConfigSetting()).be.true();
+		});
+	});
+
+	describe('debounce', () => {
+		class DebounceTest {
+			public fnCalled = 0;
+			public getterCalled = 0;
+
+			@utils.debounce(100)
+			fn(): void {
+				this.fnCalled++;
+			}
+
+			@utils.debounce(100)
+			get getter(): number {
+				this.getterCalled++;
+				return -1;
+			}
+		}
+
+		it('decorates function correctly', async () => {
+			const debounceTestObj = new DebounceTest();
+			debounceTestObj.fn();
+			debounceTestObj.fn();
+			await sleep(500);
```

Comment:
we have a few hundred ms here to play with, but this would be the only thing i'm a bit worried about reliability wise for these tests. hopefully me just being paranoid 😄

Reply:
Yeah - I don't really know of a more consistent way to test this though. I figure we can put it in and if it isn't stable I'll investigate for better options then.